### PR TITLE
integration tests: add subpackage with integration tests

### PR DESCRIPTION
**Pull request by teg**
The overall idea is to ship integration tests as a subpackage to be installed and run on the target machine, not requiring any additional dependencies or framework to be installed on the target, and allowing the test suite to be triggered form the outside.

See individual commits for details.

### CODE REVIEW

**Comment by ondrejbudai** [comment]
`673ad7d6` - Imho we should merge this asap (also it's kinda unrelated), can you cherrypick it on a different branch.

File: golang-github-osbuild-composer.spec [hunk 1]
```diff
@@ -100,16 +108,17 @@ export GOPATH=$PWD/_build:%{gopath}
 %license LICENSE
 %doc README.md
 %{_libexecdir}/osbuild-composer/
+%{_datadir}/osbuild-composer/
 %{_unitdir}/*.{service,socket}
 %{_sysusersdir}/osbuild-composer.conf
 
-%changelog
```
Comment (ondrejbudai) [inline golang-github-osbuild-composer.spec]: Please, don't kill the changelog. :)

File: internal/rpmmd/repository.go [hunk 2]
```diff
@@ -106,6 +106,34 @@ func (err *DNFError) Error() string {
...
+	path := "/repositories/" + distro + ".json"
+
+	for _, confPath := range confPaths {
+		f, err = os.Open(confPath + path)
+		if os.IsNotExist(err) {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+	}
```
Comment (larskarlitski) [inline internal/rpmmd/repository.go]: I think this loop is wrong. It doesn't stop when it finds the first repo file. Did you mean to add an uncondontional `break` at the end?

File: repositories/fedora-30.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+    "x86_64": [
+        {
+            "id": "fedora",
+            "name": "Fedora 30",
```
Comment (larskarlitski) [inline repositories/fedora-30.json]: Unrelated to this patch, but why are we including the `name` everywhere? It's not really used, is it?

File: cmd/osbuild-tests/main.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+	err := json.Unmarshal(rawReply, &reply)
+	if err != nil {
+		log.Fatalf("Unexpected reply: " + err.Error())
+	}
+	if reply.Status != true {
+		log.Fatalf("Unexpected status %v", reply.Status)
+	}
+}
+
+func runCommand(quiet bool, command ...string) json.RawMessage {
```
Comment (larskarlitski) [inline cmd/osbuild-tests/main.go]: Do you mind naming this `runComposerCLI` or something more descriptive?

File: cmd/osbuild-tests/main.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+	command = append([]string{"--json"}, command...)
```
Comment (larskarlitski) [inline cmd/osbuild-tests/main.go]: @bcl mentioned that `--json` doesn't apply to all commands and is not guaranteed to stay the same. We should probably clarify before depending on it to heavily.

File: cmd/osbuild-tests/main.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+	contents, err := ioutil.ReadAll(stdout)
+	if err != nil {
+		log.Fatalf("Could not read stdout from command: " + err.Error())
+	}
+
+	if len(contents) != 0 {
+		err = json.Unmarshal(contents, &result)
+		if err != nil {
+			// We did not get JSON, try interpreting it as TOML
```
Comment (larskarlitski) [inline cmd/osbuild-tests/main.go]: There are some commands that output neither `json` nor `toml`, like `status show`. We might need to make the output format an explicit parameter to this function at some point (no need to do it now, though).

File: golang-github-osbuild-composer.spec (hunk 1 above)
Comment (larskarlitski) [reply golang-github-osbuild-composer.spec]: Hehe :) What's the use of maintaining a changelog upstream if most entries are `New upstream release`?

File: internal/distro/distro.go [hunk 7]
```diff
@@ -40,25 +40,37 @@ type Distro interface {
...
-var registered map[string]Distro
+type DistroRegister struct {
+	distros map[string]Distro
+}
 
-func init() {
-	registered = map[string]Distro{
-		fedora30.Name: fedora30.New(),
-		rhel82.Name:   rhel82.New(),
+func NewDistroRegister(confPaths []string) *DistroRegister {
```
Comment (larskarlitski) [inline internal/distro/distro.go]: I really prefer `NewRegistry` (`Register` → `Registry` and no stuttering), but really don't worry changing it everywhere if that's too much hassle. It's just a preference, after all.

File: internal/distro/distro_test.go [hunk 8]
```diff
@@ -42,7 +42,8 @@ func TestDistro_Pipeline(t *testing.T) {
 			continue
 		}
 		t.Run(tt.Compose.OutputFormat, func(t *testing.T) {
-			d := distro.New(tt.Compose.Distro)
+			distros := distro.NewDistroRegister([]string{"../.."})
+			d := distros.GetDistro(tt.Compose.Distro)
```
Comment (larskarlitski) [inline internal/distro/distro_test.go]: We're now creating the distro objects every time they're being accessed. I think that's idempotent right now, so we are fine, but we might want to move the register object into something that's passed around.

File: golang-github-osbuild-composer.spec (hunk 1 above)
Comment (msehnout) [reply golang-github-osbuild-composer.spec]: Please kill the changelog :D

**Comment by msehnout** [comment]
I like this PR.

One question though: I'd like to see the tests as part of our CI, do you have any plan for this?

File: repositories/fedora-30.json (hunk 3 above)
Comment (teg) [reply repositories/fedora-30.json]: Yeah, I was going to ask you the same question ;) Should look into that (separately).

File: cmd/osbuild-tests/main.go (hunk 4 above)
Comment (teg) [reply cmd/osbuild-tests/main.go]: Good point! Done.

File: cmd/osbuild-tests/main.go (hunk 6 above)
Comment (teg) [reply cmd/osbuild-tests/main.go]: Yeah, I thought I'd start off trying to make everything JSON and generalize if and when we run into issues. I expect this needs to be refactored eventually, but seems to get us pretty far for now.

File: internal/distro/distro.go (hunk 7 above)
Comment (teg) [reply internal/distro/distro.go]: Ah, you are totally right. Will fix.

File: internal/distro/distro_test.go (hunk 8 above)
Comment (teg) [reply internal/distro/distro_test.go]: Agreed, for now this tries to minimize the changes, but should be followed by a clean-up.

**Comment by teg** [comment]
@msehnout yeah, the aim is to get this into CI. I'm not 100% sure how to hook it all up. I have some local scripts that do it how I imagine it should be as a proof of concept, but still need to figure out how to make that a reality. Will be the next step :)

File: cmd/osbuild-tests/main.go (hunk 5 above)
Comment (teg) [reply cmd/osbuild-tests/main.go]: Yeah, see below.

File: internal/rpmmd/repository.go (hunk 2 above)
Comment (teg) [reply internal/rpmmd/repository.go]: Ehhh... Right. I want to say I tested this, but obviously I can't have.

**Comment by teg** [comment]
Addressed the comments, except for the changelog, that's still to be resolved.

**Comment by teg** [comment]
Reverted the changelog. We can discuss that separately.

File: cmd/osbuild-tests/main.go (hunk 6 above)
Comment (larskarlitski) [reply cmd/osbuild-tests/main.go]: Yeah. This was (as most of these comments) just me thinking out loud :wink:

**Pull request merged**